### ISSUE TITLE
Result set quantity does not equal MySQL rows quantity

### ISSUE THREAD

**Issue by Lodin** [mysql-d/mysql-native#56]
Hello. I'm trying to make simple site based on vibe.d and MySQL. And I faced with some problems with MySQL results. In short, with some conditions there are less rows in my result than exists in DB.

There is SQL dump to reproduce the bug
```sql
CREATE DATABASE  IF NOT EXISTS `mydb` /*!40100 DEFAULT CHARACTER SET latin1 */;
USE `mydb`;
-- MySQL dump 10.13  Distrib 5.5.41, for debian-linux-gnu (x86_64)
--
-- Host: 127.0.0.1    Database: mydb
-- ------------------------------------------------------
-- Server version   5.5.41-0ubuntu0.14.04.1

/*!40101 SET @OLD_CHARACTER_SET_CLIENT=@@CHARACTER_SET_CLIENT */;
/*!40101 SET @OLD_CHARACTER_SET_RESULTS=@@CHARACTER_SET_RESULTS */;
/*!40101 SET @OLD_COLLATION_CONNECTION=@@COLLATION_CONNECTION */;
/*!40101 SET NAMES utf8 */;
/*!40103 SET @OLD_TIME_ZONE=@@TIME_ZONE */;
/*!40103 SET TIME_ZONE='+00:00' */;
/*!40014 SET @OLD_UNIQUE_CHECKS=@@UNIQUE_CHECKS, UNIQUE_CHECKS=0 */;
/*!40014 SET @OLD_FOREIGN_KEY_CHECKS=@@FOREIGN_KEY_CHECKS, FOREIGN_KEY_CHECKS=0 */;
/*!40101 SET @OLD_SQL_MODE=@@SQL_MODE, SQL_MODE='NO_AUTO_VALUE_ON_ZERO' */;
/*!40111 SET @OLD_SQL_NOTES=@@SQL_NOTES, SQL_NOTES=0 */;

--
-- Table structure for table `user`
--

DROP TABLE IF EXISTS `user`;
/*!40101 SET @saved_cs_client     = @@character_set_client */;
/*!40101 SET character_set_client = utf8 */;
CREATE TABLE `user` (
  `id` int(11) NOT NULL AUTO_INCREMENT,
  `name` varchar(255) NOT NULL,
  `registered_at` datetime DEFAULT NULL,
  `updated_at` datetime DEFAULT NULL,
  `last_visit` datetime DEFAULT NULL,
  `removed_at` datetime DEFAULT NULL,
  `rating_quantity` int(11) NOT NULL,
  PRIMARY KEY (`id`)
) ENGINE=MyISAM AUTO_INCREMENT=11 DEFAULT CHARSET=utf8;
/*!40101 SET character_set_client = @saved_cs_client */;

--
-- Dumping data for table `user`
--

LOCK TABLES `user` WRITE;
/*!40000 ALTER TABLE `user` DISABLE KEYS */;
INSERT INTO `user` VALUES
(1,'admin','2014-12-01 00:45:46',NULL,'2014-12-01 00:00:00',NULL,0),
(2,'testuser1','2014-12-01 12:08:34',NULL,'2014-12-01 00:00:00',NULL,0),
(3,'test2','2014-12-01 12:08:34',NULL,'2014-12-01 00:00:00',NULL,0),
(4,'test3','2014-12-01 12:08:34',NULL,'2014-12-01 00:00:00',NULL,0),
(5,'test4','2014-12-01 12:08:34',NULL,'2014-12-01 00:00:00',NULL,0);
/*!40000 ALTER TABLE `user` ENABLE KEYS */;
UNLOCK TABLES;
/*!40103 SET TIME_ZONE=@OLD_TIME_ZONE */;

/*!40101 SET SQL_MODE=@OLD_SQL_MODE */;
/*!40014 SET FOREIGN_KEY_CHECKS=@OLD_FOREIGN_KEY_CHECKS */;
/*!40014 SET UNIQUE_CHECKS=@OLD_UNIQUE_CHECKS */;
/*!40101 SET CHARACTER_SET_CLIENT=@OLD_CHARACTER_SET_CLIENT */;
/*!40101 SET CHARACTER_SET_RESULTS=@OLD_CHARACTER_SET_RESULTS */;
/*!40101 SET COLLATION_CONNECTION=@OLD_COLLATION_CONNECTION */;
/*!40111 SET SQL_NOTES=@OLD_SQL_NOTES */;

-- Dump completed on 2015-02-01 16:00:30
```
As you can see, there should be `5` rows in result set. 

There is the code I use to connect to MySQL
```d
import std.stdio;
import mysql.common;
import mysql.connection;
import mysql.result;

static this()
{
	auto conn = new Connection("host=localhost;port=3306;user=root;pwd=;db=mydb");
	auto res = conn.query("SELECT _t1.id, _t1.name, _t1.rating_quantity, _t1.registered_at, _t1.updated_at, _t1.last_visit, _t1.removed_at FROM user AS _t1 ORDER BY _t1.id");
	while(!res.empty)
	{
		writefln("ID: %s, Name: %s", res.front[0], res.front[1]);
		res.popFront();
	}
}

ResultSet query()(Connection cn, string sql)
{
	auto cmd = cn.prepare(sql);
	return cmd.query();
}

Command prepare(Connection cn, string sql)
{
	auto cmd = Command(cn);
	cmd.sql = sql;
	cmd.prepare();
	return cmd;
}

ResultSet query()(Command cmd)
{
	return cmd.execPreparedResult();
}
```
And there my results:
```
ID: 1, Name: admin
ID: 3, Name: test2
ID: 5, Name: test4
```
If I change next conditions, everything returns to usual:
* Query. If I place `_t1.rating_quantity` to the end of query (like it is in DB table), I can get all results.
* Date of `last_visit`. If it does not have `00:00:00` as time of day, I can get that result.

If I do that, I will get next results:
```
ID: 1, Name: admin
ID: 2, Name: testuser1
ID: 3, Name: test2
ID: 4, Name: test3
ID: 5, Name: test4
```

Maybe it could get some tips, so I'll place it here (comments is mine):
```d
// Method I explored to find error
ResultSet execPreparedResult(ColumnSpecialization[] csa = null)
{
    ulong ra;
    enforceEx!MYX(execPrepared(ra), "The executed query did not produce a result set.");
    uint alloc = 20;
    Row[] rra;
    rra.length = alloc;
    uint cr = 0;
    _rsh = ResultSetHeaders(_con, _fieldCount);
    if (csa !is null)
        _rsh.addSpecializations(csa);
    _headersPending = false;
    ubyte[] packet;
    for (size_t i = 0;; i++)
    {
        packet = _con.getPacket();
        
        // If everything is OK and I'm getting all results, query ends here
        if (packet.isEOFPacket())
            break;

        Row row = Row(_con, packet, _rsh, true);
        if (cr >= alloc)
        {
            alloc = (alloc*3)/2;
            rra.length = alloc;
        }
        rra[cr++] = row;

        // If I'm getting wrong set of result, query ends here
        if (!packet.empty && packet.isEOFPacket())
            break;
    }
    _rowsPending = _pendingBinary = false;
    rra.length = cr;
    ResultSet rs = ResultSet(rra, _rsh.fieldNames);
    return rs;
}
```

**Comment by Temtaime** [comment]
Hi ! I also faced this problem. Is there's no solution ?

**Comment by Temtaime** [comment]
@s-ludwig @simendsjo any fix ?

**Comment by Temtaime** [comment]
Sadly that there's no reply.
I have to move to another library then. I have a big system with complicated queries and i cannot divide them, so missing rows makes me sad.

**Comment by simendsjo** [comment]
I haven't looked at this project in ages, but I reduced the case further. It looks like there is some issue using prepared statements together with datetimes. Seems like it skips every other row. If I try to get other fields the bug might disappear.

I haven't seen anything strange in `executePreparedResult`, so my guess is `consumeBinaryValueIfComplete` or something in the packet helpers.

Here's some reduces code:

`mysqln_issue_56.d`:
```d
import std.stdio;
import mysql.common, mysql.connection, mysql.result;

void main()
{
    auto conn = new Connection("host=localhost;port=3306;user=root;pwd=;db=mysqln_issue_56");

    auto cmd = Command(conn);
    cmd.sql = "SELECT a FROM t";
    cmd.prepare();

    auto res = cmd.execPreparedResult();
    while(!res.empty)
    {
        writefln("Value: %s", res.front[0]);
        res.popFront();
    }
}
```

`mysqln_issue_56.sql`:
```sql
CREATE DATABASE IF NOT EXISTS mysqln_issue_56;
USE mysqln_issue_56;
DROP TABLE IF EXISTS t;
CREATE TABLE t ( a datetime DEFAULT NULL);
INSERT INTO t VALUES
   ('2015-03-28 00:00:00')
  ,('2015-03-29 00:00:00')
  ,('2015-03-31 00:00:00')
  ,('2015-03-31 00:00:00')
  ,('2015-03-31 00:00:00')
  ,('2015-03-31 00:00:00')
  ,('2015-04-01 00:00:00')
  ,('2015-04-02 00:00:00')
  ,('2015-04-03 00:00:00')
  ,('2015-04-04 00:00:00')
;
```

**Comment by Temtaime** [comment]
There's problem not only with datetimes. I have no datetimes and rows seems to disappear anyway.
OK, thanks, i'll try to look.

**Comment by simendsjo** [comment]
Ok, but it's probably related to prepared statements which has a quite hairy implementation. Try reducing the rows to only two equal dates and step through the `packet.d` and `packet_helpers.d` functions. I don't know who's maintaining this project anymore. @Abscissa? @sshamov?

**Comment by Abscissa** [comment]
I'm looking into it right now. It does appear that this problem only occurs for prepared statements, so in the meantime, you could switch to non-prepared as a workaround:

From:
```
cmd.sql = "SELECT...";
cmd.prepare();
auto res = cmd.execPreparedResult();
```

To:
```
cmd.sql = "SELECT...";
auto res = cmd.execSQLResult();
```

**Comment by Abscissa** [comment]
Fixed version is tagged v0.1.2
